### PR TITLE
Reference assemblies reachable from current module only

### PR DESCRIPTION
**Pull request by cston**
### Customer scenario

Evaluating expressions in the Locals, Watch, or Callstack windows is slow when the process contains a large number of modules and not all modules are referenced by the current module.

### Bugs this fixes

591114

### Workarounds, if any

None

### Risk

Medium risk. The user must opt-in to the fix via a registry setting, but there are some changes required to maintain the existing behavior alongside the fix.

### Performance impact

No performance impact unless opted-in.

### Is this a regression from a previous update?

No.

### How was the bug found?

Customer reported.

### CODE REVIEW

**Comment by cston** [comment]
cc @jinujoseph, @tmat

**Comment by cston** [comment]
@tmat, @ivanbasov please review.
cc @jinujoseph

File: src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/CSharpExpressionCompiler.cs [hunk 1]
```diff
@@ -89,33 +147,49 @@ internal override DkmCompilerId CompilerId
                     localSignatureToken);
             }
 
-            var previous = appDomain.GetMetadataContext<CSharpMetadataContext>();
+            var previous = getMetadataContext(appDomain);
+            if (!previous.Matches(metadataBlocks))
+            {
+                previous = null;
+            }
+            if (previous != null && previous.ModuleVersionId != moduleVersionId)
```
Comment (cston) [inline src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/CSharpExpressionCompiler.cs]: If the fix is disabled (if `kind == AllAssemblies`), we should not be setting or checking `previous.ModuleVersionId`.

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/EEMetadataReferenceResolver.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        public override PortableExecutableReference ResolveMissingAssembly(MetadataReference definition, AssemblyIdentity referenceIdentity)
+        {
+#if DEBUG
+            int n;
+            Requests.TryGetValue(referenceIdentity, out n);
+            Requests[referenceIdentity] = n + 1;
+#endif
+            MetadataReference reference;
+            _referencesByIdentity.TryGetValue(referenceIdentity, out reference);
+            return (PortableExecutableReference)reference;
```
Comment (cston) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/EEMetadataReferenceResolver.cs]: Should return highest version of assembly if exact version is missing and there is a higher version available.

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs [hunk 3]
```diff
@@ -105,42 +114,98 @@ internal static partial class MetadataUtilities
...
                 }
-                var reference = MakeAssemblyMetadata(metadata, modulesByName);
-                referencesBuilder.Add(reference);
+            }
+
+            if (kind == MakeAssemblyReferencesKind.AllReferences)
+            {
+                Debug.Assert(referencesBuilder.Count == 0);
+                // CommonReferenceManager<TCompilation, TAssemblySymbol>.Bind()
+                // expects COR library to be included in the explicit assemblies (see bug #...).
```
Comment (ivanbasov) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs]: >#... [](start = 90, length = 4)

Would you like to specify the bug #?

File: src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/ExpressionCompilerTestBase.cs [hunk 4]
```diff
@@ -111,17 +169,31 @@ internal RuntimeInstance CreateRuntimeInstance(IEnumerable<ModuleInstance> modul
...
+        internal static void GetContextState(
+            ImmutableArray<ModuleInstance> moduleInstances,
+            string methodOrTypeName,
+            out ImmutableArray<MetadataBlock> blocks,
+            out Guid moduleVersionId,
+            out ISymUnmanagedReader symReader,
+            out int methodOrTypeToken,
+            out int localSignatureToken,
+            Guid moduleVersionIdOpt = default,
+            MakeAssemblyReferencesKind kind = MakeAssemblyReferencesKind.AllAssemblies) // TODO: Revert this change. No need for MakeAssemblyReferencesKind here.
```
Comment (ivanbasov) [inline src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/ExpressionCompilerTestBase.cs]: >TODO: Revert this change. [](start = 91, length = 25)

Is it for now or a bug should be created for this TODO?

File: src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/ReferencedModulesTests.cs [hunk 5]
```diff
@@ -30,6 +30,378 @@ public class ReferencedModulesTests : ExpressionCompilerTestBase
...
+                methodData.VerifyIL(
+@"{
+// Code size        6 (0x6)
+.maxstack  1
+IL_0000:  newobj     ""B2..ctor()""
+IL_0005:  ret
+}");
+                VerifyResolutionRequests(context, (identityA1, identityA1, 1));
+
+                // TODO: Test { A1.M, B1.M, C.M } with all assemblies.
```
Comment (ivanbasov) [inline src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/ReferencedModulesTests.cs]: >TODO [](start = 19, length = 4)

Should this test be added now?

File: src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/ReferencedModulesTests.cs [hunk 6]
```diff
@@ -30,6 +30,378 @@ public class ReferencedModulesTests : ExpressionCompilerTestBase
...
+        {
+            var assemblyContexts = appDomainContext.AssemblyContexts;
+            return assemblyContexts != null && assemblyContexts.TryGetValue(default(MetadataContextId), out CSharpMetadataContext context) ?
+                context :
+                default;
+        }
+
+        private static void VerifyAppDomainMetadataContext(AppDomain appDomain)
+        {
+            // TODO: Verify set of AssemblyMetadataContext instances.
```
Comment (ivanbasov) [inline src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/ReferencedModulesTests.cs]: >TODO [](start = 15, length = 4)

?

**Comment by cston** [comment]
cc @jaredpar

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/ExpressionEvaluatorFatalError.cs [hunk 7]
```diff
@@ -53,7 +47,24 @@ static ExpressionEvaluatorFatalError()
             {
                 Debug.Assert(false, "Failure checking registry key: " + ex.ToString());
             }
+            return null;
+        }
+
+        internal static bool GetBoolRegistryValue(string name)
+        {
+            var value = RegistryHelpers.GetRegistryValue(name);
+            if ((value != null) && (value is int))
```
Comment (jaredpar) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/ExpressionEvaluatorFatalError.cs]: >if ((value != null) && (value is int)) [](start = 12, length = 38)

Isn't the null check redundant here?

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs (hunk 3 above)
Comment (cston) [reply src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs]: Removed `see bug #`.

**Comment by cston** [comment]
@tmat, please review.
@ivanbasov, please take a second look since there are a number of changes (caching in particular) since your review, thanks.

**Comment by ivanbasov** [comment]
reviewed once again. LGTM. Thanks!

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/EEMetadataReferenceResolver.cs [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+    internal sealed class EEMetadataReferenceResolver : MetadataReferenceResolver
+    {
+        private readonly AssemblyIdentityComparer _identityComparer;
+        private readonly Dictionary<string, ImmutableArray<(AssemblyIdentity, MetadataReference)>> _referencesByIdentity;
+
+#if DEBUG
+        internal readonly Dictionary<AssemblyIdentity, (AssemblyIdentity Identity, int Count)> Requests = new Dictionary<AssemblyIdentity, (AssemblyIdentity Identity, int Count)>();
+#endif
+
+        internal EEMetadataReferenceResolver(AssemblyIdentityComparer identityComparer, Dictionary<string, ImmutableArray<(AssemblyIdentity, MetadataReference)>> referencesByIdentity)
```
Comment (tmat) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/EEMetadataReferenceResolver.cs]: >Dictionary [](start = 88, length = 10)

IReadOnlyDictionary

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/EEMetadataReferenceResolver.cs [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+            _referencesByIdentity = referencesByIdentity;
+        }
+
+        public override bool ResolveMissingAssemblies => true;
+
+        public override PortableExecutableReference ResolveMissingAssembly(MetadataReference definition, AssemblyIdentity referenceIdentity)
+        {
+            ImmutableArray<(AssemblyIdentity, MetadataReference)> references;
+            (AssemblyIdentity, MetadataReference) result = default;
+            if (_referencesByIdentity.TryGetValue(referenceIdentity.Name, out references))
```
Comment (tmat) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/EEMetadataReferenceResolver.cs]: >references [](start = 78, length = 10)

nit: `out var references`

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/EEMetadataReferenceResolver.cs [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+            _identityComparer = identityComparer;
+            _referencesByIdentity = referencesByIdentity;
+        }
+
+        public override bool ResolveMissingAssemblies => true;
+
+        public override PortableExecutableReference ResolveMissingAssembly(MetadataReference definition, AssemblyIdentity referenceIdentity)
+        {
+            ImmutableArray<(AssemblyIdentity, MetadataReference)> references;
+            (AssemblyIdentity, MetadataReference) result = default;
```
Comment (tmat) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/EEMetadataReferenceResolver.cs]: >AssemblyIdentity, MetadataReference [](start = 13, length = 35)

Perhaps we could name these so that we don't need to wonder what `Item1` and `Item2` are below

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/EEMetadataReferenceResolver.cs [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+            if (_referencesByIdentity.TryGetValue(referenceIdentity.Name, out references))
+            {
+                result = GetBestMatch(references, referenceIdentity);
+            }
+#if DEBUG
+            (AssemblyIdentity Identity, int Count) request;
+            if (!Requests.TryGetValue(referenceIdentity, out request))
```
Comment (tmat) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/EEMetadataReferenceResolver.cs]: >out r [](start = 57, length = 5)

out var?

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/EEMetadataReferenceResolver.cs [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+        {
+            throw ExceptionUtilities.Unreachable;
+        }
+
+        public override int GetHashCode()
+        {
+            throw ExceptionUtilities.Unreachable;
+        }
+
+        private (AssemblyIdentity, MetadataReference) GetBestMatch(ImmutableArray<(AssemblyIdentity, MetadataReference)> references, AssemblyIdentity referenceIdentity)
```
Comment (tmat) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/EEMetadataReferenceResolver.cs]: >(AssemblyIdentity, MetadataReference) [](start = 82, length = 37)

Perhaps we could name these so that we don't need to wonder what Item1 and Item2 are below

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/InstructionDecoder.cs [hunk 13]
```diff
@@ -22,6 +20,18 @@ internal abstract class InstructionDecoder<TCompilation, TMethodSymbol, TModuleS
             memberOptions: SymbolDisplayMemberOptions.IncludeContainingType | SymbolDisplayMemberOptions.IncludeExplicitInterface,
             miscellaneousOptions: SymbolDisplayMiscellaneousOptions.UseSpecialTypes);
 
+        private readonly bool _useReferencedAssembliesOnly;
+
+        internal InstructionDecoder()
+        {
+            _useReferencedAssembliesOnly = ExpressionCompiler.GetUseReferencedAssembliesOnlySetting();
```
Comment (tmat) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/InstructionDecoder.cs]: >ExpressionCompiler.GetUseReferencedAssembliesOnlySetting(); [](start = 43, length = 59)

Shouldn't this rather flow from the ExpressionCompiler instance?

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataContext.cs [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+using System.Collections.Immutable;
+using System.Linq;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.ExpressionEvaluator
+{
+    // Wrapper around Guid to ensure callers have asked for the correct id
+    // rather than simply using the ModuleVersionId (which is unnecessary
+    // when the Compilation references all loaded assemblies).
+    internal struct MetadataContextId : IEquatable<MetadataContextId>
```
Comment (tmat) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataContext.cs]: > s [](start = 12, length = 2)

readonly struct

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataContext.cs [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+                    return default;
+                case MakeAssemblyReferencesKind.AllReferences:
+                    return new MetadataContextId(moduleVersionId);
+                default:
+                    throw ExceptionUtilities.UnexpectedValue(kind);
+            }
+        }
+    }
+
+    internal struct MetadataContext<TAssemblyContext>
```
Comment (tmat) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataContext.cs]: >  [](start = 12, length = 1)

readonly struct

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataContext.cs [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+            {
+                case MakeAssemblyReferencesKind.AllAssemblies:
+                    return default;
+                case MakeAssemblyReferencesKind.AllReferences:
+                    return new MetadataContextId(moduleVersionId);
+                default:
+                    throw ExceptionUtilities.UnexpectedValue(kind);
+            }
+        }
+    }
```
Comment (tmat) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataContext.cs]: Move to a separate file

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs [hunk 17]
```diff
@@ -9,13 +9,19 @@
...
 using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.ExpressionEvaluator
 {
+    internal enum MakeAssemblyReferencesKind
+    {
+        AllAssemblies,
+        AllReferences,
+        DirectReferencesOnly,
+    }
```
Comment (tmat) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs]: Move to separate file

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs [hunk 18]
```diff
@@ -105,42 +114,98 @@ internal static partial class MetadataUtilities
                 modulesByName[name] = modulesByName.ContainsKey(name) ? null : metadata;
             }
 
+            // We don't walk the references of winmd assemblies currently. (That would require
+            // walking references from the current module and also the winmd assemblies.)
+            // So if there are any winmd assemblies, we'll use all assemblies.
```
Comment (tmat) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs]: >So if there are any winmd assemblies, we'll use all assemblies. [](start = 15, length = 63)

Should we have an issue for this? We might want to address this in 15.8

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs [hunk 19]
```diff
@@ -105,42 +114,98 @@ internal static partial class MetadataUtilities
...
+            if (kind == MakeAssemblyReferencesKind.AllReferences && runtimeWinMdBuilder.Any())
+            {
+                kind = MakeAssemblyReferencesKind.AllAssemblies;
+            }
+
             // Build assembly references from modules in primary module manifests.
             var referencesBuilder = ArrayBuilder<MetadataReference>.GetInstance();
-            var identitiesBuilder = (identityComparer == null) ? null : ArrayBuilder<AssemblyIdentity>.GetInstance();
+            var identitiesBuilder = (kind == MakeAssemblyReferencesKind.DirectReferencesOnly) ? ArrayBuilder<AssemblyIdentity>.GetInstance() : null;
+            referencesByIdentity = (kind == MakeAssemblyReferencesKind.AllReferences) ? new Dictionary<string, ImmutableArray<(AssemblyIdentity, MetadataReference)>>(StringComparer.OrdinalIgnoreCase) : null;
```
Comment (tmat) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs]: >referencesByIdentity [](start = 12, length = 20)

better name: `referencesBySimpleName`

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs [hunk 20]
```diff
@@ -105,42 +114,98 @@ internal static partial class MetadataUtilities
...
+
+                var reference = MakeAssemblyMetadata(metadata, modulesByName);
+
+                if (kind == MakeAssemblyReferencesKind.AllReferences)
+                {
+                    var reader = metadata.MetadataReader;
+                    var identity = reader.ReadAssemblyIdentityOrThrow();
+                    if (referencesByIdentity != null)
+                    {
+                        string identityKey = identity.Name;
```
Comment (tmat) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs]: >identityKey [](start = 31, length = 11)

`simpleName`

Comment (tmat) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs]: I'd inline this variable

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs [hunk 21]
```diff
@@ -105,42 +114,98 @@ internal static partial class MetadataUtilities
...
+                        ImmutableArray<(AssemblyIdentity, MetadataReference)> references;
+                        if (!referencesByIdentity.TryGetValue(identityKey, out references))
```
Comment (tmat) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs]: >out [](start = 75, length = 3)

out var

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs [hunk 22]
```diff
@@ -105,42 +114,98 @@ internal static partial class MetadataUtilities
...
                     continue;
                 }
+
+                var reference = MakeAssemblyMetadata(metadata, modulesByName);
+
+                if (kind == MakeAssemblyReferencesKind.AllReferences)
+                {
+                    var reader = metadata.MetadataReader;
+                    var identity = reader.ReadAssemblyIdentityOrThrow();
+                    if (referencesByIdentity != null)
```
Comment (tmat) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs]: >(referencesByIdentity != null [](start = 23, length = 29)

Isn't this always true?

**Comment by tmat** [comment]
>         private static PortableExecutableReference MakeAssemblyMetadata(ModuleMetadata metadata, Dictionary<string, ModuleMetadata> modulesByName)

Rename to `MakeAssemblyReference`?

---
Refers to: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs:315 in 54ecd0f. [](commit_id = 54ecd0fb0bb338c5265905d197b8f1fa775759ab, deletion_comment = False)

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs [hunk 23]
```diff
@@ -105,42 +114,98 @@ internal static partial class MetadataUtilities
...
+                if (corLibrary != null && referencesByIdentity.TryGetValue(corLibrary.Name, out var corLibraryReferences))
+                {
+                    referencesBuilder.Add(corLibraryReferences[0].Item2);
+                }
+                Debug.Assert(targetReference != null);
+                if (targetReference != null)
+                {
+                    referencesBuilder.Add(targetReference);
+                }
             }
```
Comment (tmat) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs]: There seems to be 2 code paths here that are interwoven but almost entirely distinct. `Kind == AllReferences` and otherwise. Consider refactoring to two different methods or blocks.

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs (hunk 20 above)
Comment (cston) [reply src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs]: It's used a couple of times. Since it's not obvious what the dictionary key is, I'll keep the variable.

---
In reply to: [181511479](https://github.com/dotnet/roslyn/pull/25955#discussion_r181511479) [](ancestors = 181511479)

File: src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/CompilationExtensions.cs [hunk 24]
```diff
@@ -83,24 +85,26 @@ internal static PEModuleSymbol GetModule(this CSharpCompilation compilation, Gui
...
-            var references = metadataBlocks.MakeAssemblyReferences(moduleVersionId, IdentityComparer);
-            return references.ToCompilation();
+            return ToCompilation(metadataBlocks, moduleVersionId, kind: MakeAssemblyReferencesKind.DirectReferencesOnly);
         }
 
-        internal static CSharpCompilation ToCompilation(this ImmutableArray<MetadataReference> references)
+        internal static CSharpCompilation ToCompilation(this ImmutableArray<MetadataBlock> metadataBlocks, Guid moduleVersionId, MakeAssemblyReferencesKind kind)
         {
+            Dictionary<string, ImmutableArray<(AssemblyIdentity, MetadataReference)>> referencesByIdentity;
+            var references = metadataBlocks.MakeAssemblyReferences(moduleVersionId, IdentityComparer, kind, out referencesByIdentity);
```
Comment (tmat) [inline src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/CompilationExtensions.cs]: >out  [](start = 108, length = 4)

`out var` to avoid declaring the var with the very long type name

Comment (tmat) [inline src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/CompilationExtensions.cs]: >referencesByIdentity [](start = 112, length = 20)

`referencesBySimpleName`

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs (hunk 20 above)
Comment (tmat) [reply src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs]: I think this would read better: `referencesBySimpleName.TryGetValue(identity.Name, out references)`

**Comment by tmat** [comment]
LGTM. Just some minor feedback.

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/InstructionDecoder.cs (hunk 13 above)
Comment (cston) [reply src/ExpressionEvaluator/Core/Source/ExpressionCompiler/InstructionDecoder.cs]: Yes, ideally. Added a comment for now.

---
In reply to: [181508668](https://github.com/dotnet/roslyn/pull/25955#discussion_r181508668) [](ancestors = 181508668)

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs (hunk 18 above)
Comment (cston) [reply src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs]: Logged #26157.

---
In reply to: [181510450](https://github.com/dotnet/roslyn/pull/25955#discussion_r181510450) [](ancestors = 181510450)

**Comment by cston** [comment]
@MattGertz for 15.7 approval.
cc @jinujoseph

**Comment by jinujoseph** [comment]
approved to merge via [link](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/591114) for 15.7.Preview5

**Comment by jinujoseph** [comment]
vendor test approval pending @richaverma1

**Comment by cston** [comment]
@dotnet-bot test windows_debug_vs-integration_prtest
